// next-connect/dist/commonjs/edge.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createEdgeRouter = exports.getPathname = exports.EdgeRouter = void 0;
const router_js_1 = require("./router.cjs");
class EdgeRouter extends router_js_1.Router {
    constructor() {
        super();
    }
    prepareRequest(req, ctx, findResult) {
        req.params = {
            ...findResult.params,
            ...req.params, // original params will take precedence
        };
    }
    async run(req, ctx) {
        const result = this.find(req.method, getPathname(req));
        if (!result.fns.length)
            return;
        this.prepareRequest(req, ctx, result);
        return router_js_1.Router.exec(result.fns, req, ctx);
    }
    handler(options = {}) {
        const onNoMatch = options.onNoMatch || onnomatch;
        const onError = options.onError || onerror;
        return async (req, ctx) => {
            const result = this.find(req.method, getPathname(req));
            this.prepareRequest(req, ctx, result);
            try {
                if (result.fns.length === 0 || result.middleOnly) {
                    return await onNoMatch(req, ctx);
                }
                else {
                    return await router_js_1.Router.exec(result.fns, req, ctx);
                }
            }
            catch (err) {
                return onError(err, req, ctx);
            }
        };
    }
}
exports.EdgeRouter = EdgeRouter;
function onnomatch(req) {
    return new Response(req.method !== "HEAD" ? `Route ${req.method} ${req.url} not found` : null, { status: 404 });
}
function onerror(err) {
    console.error(err);
    return new Response("Internal Server Error", { status: 500 });
}
function getPathname(req) {
    return (req.nextUrl || new URL(req.url)).pathname;
}
exports.getPathname = getPathname;
function createEdgeRouter() {
    return new EdgeRouter();
}
exports.createEdgeRouter = createEdgeRouter;
