// next-connect/dist/commonjs/node.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRouter = exports.getPathname = exports.NodeRouter = void 0;
const router_js_1 = require("./router.cjs");
class NodeRouter extends router_js_1.Router {
    constructor() {
        super();
    }
    prepareRequest(req, res, findResult) {
        req.params = {
            ...findResult.params,
            ...req.params, // original params will take precedence
        };
    }
    async run(req, res) {
        const result = this.find(req.method, getPathname(req.url));
        if (!result.fns.length)
            return;
        this.prepareRequest(req, res, result);
        return router_js_1.Router.exec(result.fns, req, res);
    }
    handler(options = {}) {
        const onNoMatch = options.onNoMatch || onnomatch;
        const onError = options.onError || onerror;
        return async (req, res) => {
            const result = this.find(req.method, getPathname(req.url));
            this.prepareRequest(req, res, result);
            try {
                if (result.fns.length === 0 || result.middleOnly) {
                    await onNoMatch(req, res);
                }
                else {
                    await router_js_1.Router.exec(result.fns, req, res);
                }
            }
            catch (err) {
                await onError(err, req, res);
            }
        };
    }
}
exports.NodeRouter = NodeRouter;
function onnomatch(req, res) {
    res.statusCode = 404;
    res.end(req.method !== "HEAD"
        ? `Route ${req.method} ${req.url} not found`
        : undefined);
}
function onerror(err, req, res) {
    res.statusCode = 500;
    console.error(err);
    res.end("Internal Server Error");
}
function getPathname(url) {
    const queryIdx = url.indexOf("?");
    return queryIdx !== -1 ? url.substring(0, queryIdx) : url;
}
exports.getPathname = getPathname;
function createRouter() {
    return new NodeRouter();
}
exports.createRouter = createRouter;
